00X-tracing.md

```diff
@@ -0,0 +1,581 @@
+   * @param enable Optional value specifying wiether recording will be enabled for the specified
+   *   category or categories. Defaults to true.
+   */
+  enableRecording(category: string | string[], enable?: boolean): void;
+
+  /**
+   * Emits a tracing event using an event object.
+   *
+   * @param e The event to be emitted. The event object must specify one or more categories.
+   * @returns Nonzero event id (either provided or generated) if the event was emitted, or null if
+   *   the event was not emitted because tracing was not enabled for any of the event categories.
+   */
+  emit(e: TracingEvent): number | null;
+
+  /**
+   * Emits a tracing event using an event object, with categories specified separately.
+   *
+   * @param category Required tracing category name or array of one or more category names for
+   *   the tracing event. Overrides any categories specified in the event object.
+   * @param e The event to be emitted.
+   * @returns Nonzero event id (either provided or generated) if the event was emitted, or null if
+   *   the event was not emitted because tracing was not enabled for any of the event categories.
+   */
+  emit(category: string | string[], e: TracingEvent): number | null;
+}
+```
+
+### 2.4 Emit API overloads
+The `emit()` method defined above mostly follows the familiar `EventEmitter` pattern for emitting
+events as objects. However some additional overloads are proposed that take separate parameters
+instead of an event object. Using these overloads, the four emitted events in the previous code
+example can be rewritten more concisely:
+
+```javascript
+tracing.emit(category, 'myEvent');
+
+const scopeId = tracing.begin(category, 'myTask');
+tracing.end(category, 'myTask', scopeId);
+
+tracing.count(category, 'myCounter', 3);
+```
+
+These overloads are slightly more efficient because they doesn't require creation of an object for
+the event. And for simple cases they are easier to write and read.
+
+```typescript
+interface Tracing {
+  /**
+   * Emits a tracing event of type "instant" using an event name and optional event args. For
+   * simple "instant" events, this is slightly more efficient than the other emit() overloads
+   * because it doesn't require creation of an object for the event.
+   *
+   * @param category Required tracing category name or array of one or more category names for
+   *   the tracing event.
+   * @param name Required name for the event.
+   * @param args Optional dictionary of name-value pairs, or a function that returns a dictionary.
+   *   If a function, it will be invoked to retrieve the args only if tracing is enabled for the
+   *   category.
+   * @returns True if the event was emitted; false if the event was not emitted because tracing
+   *   was not enabled for any of the event categories.
+   */
+  emit(
+    category: string | string[],
+    name: string,
+    args?: { [name: string]: any } | () => { [name: string]: any }): boolean;
```

Comment:
+1 for `instant` because it matches the terms used in the trace format descriptions. It makes it easier to find it elsewhere.